src/content/en/2021/caching.md

```diff
@@ -17,4 +20,414 @@ featured_stat_3: TODO
+  sheets_gid="1944529311",
+  sql_file="cache_control_directives.sql"
+) }}
+
+Cache Control directives, 61% of mobile requests include a Cache-Control response header with a `max-age` directive.
+
+The most misused Cache Control directive is `set-cookie`, used for 0.07% of total directives for Desktop and 0.08% for Mobile.
+
+Fun fact the the largest max-age value is 1625540701883604800000 seconds or 51545557517871.8 years (51 trillion).
+
+{{ figure_markup(
+  caption="TODO",
+  content="51 trillion years",
+  classes="big-number",
+  sheets_gid="529870849",
+  sql_file="todo.sql"
+) }}
+
+Talk about `immutable` directive, 12% of desktop requests containing Cache-Control: immutable are on the "static.parastorage.com" host.
+
+### `304` Not Modified status
+
+{{ figure_markup(
+  image="http-304-by-caching-strategy.png",
+  caption="HTTP 304 response rate by caching strategy.",
+  description="TODO",
+  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vSGgVDZ9RkFQLmk5C3_siIcH-8macUEZMobcC0o1z8frYj8NOkI_C2s_yE5ppMdxDAD5INjNsCBa3h1/pubchart?oid=1965033294&format=interactive",
+  sheets_gid="1136058277",
+  sql_file="valid_if_none_match_returns_304.sql"
+) }}
+
+### Validity of date strings
+
+Talk about invalid date, last modified and expires, (0.11% of dates in desktop requests are invalid).
+
+Desktop - Valid date strings:
+99.18% using date, 72.57% using last-modified, 55.48% using expires
+
+Mobile - Valid date strings:
+99.07% using date, 70.55% using last-modified, 55.57% using expires
+
+Although 99% of both Desktop and Mobile pages use valid date strings, only 55% of Desktop and 56% of Mobile pages use 'expires'. Last-modified is used at a higher percentage, at 73% for Desktop and 71% for Mobile. While the amount of invalid 'date' and 'last-modified' directives is less than 1%, 2.75% of Mobile and 3.20% of Desktop 'expires' are being used incorrectly.
+
+What is or would be the impact from using expires more effectively?
+
+Could we possibly see big wins from getting more consistent here?
+
+### Vary
+
+An essential step in caching a resource is understanding if it was already previously cached. The browser usually uses the url as the cache key. At the same time, requests with similar urls but different `accept-encoding` could be cached incorrectly. That's why we use the Vary header to instruct the browser to add a value of one or more headers to the cache key.
+
+The most popular Vary header is `accept-encoding` with 90.3% usage, followed by `user-agent` with 10.9%, `origin` with 10.1%, and `accept` with 4.8%. 
+
+[TODO Add figure showing Vary header adoption across all requests (pct_using_vary) and cache-control (pct_of_vary_with_cache_control), data from https://docs.google.com/spreadsheets/d/1-v3yR0LZIC3t4zWtqTgR3jJsKjjRMP-HATU2caP8e2c/edit#gid=1033782866 and https://docs.google.com/spreadsheets/d/1fYmpSN3diOiFrscS75NsjfsrKXzxxhUMNcYSqXnQJQU/edit#gid=1115686547 Potentiall we mighe want to show 2020 vs 2021]
```

Comment:
cc @rviscomi

Reply:
Could you clarify what you want the chart to look like?

Not sure if pct_using_vary and pct_of_vary_with_cache_control are comparable since the first is over all requests and the second is only over requests that use Vary. For those I think "big number" figures may be more appropriate.

Are you looking for something that shows the different vary_header values in terms of pct_of_vary? (visualizing the sentence above)

Reply:
Ping

Reply:
Give me 45 minutes and I will revisit this part again. Thx for the patience.

Reply:
@rviscomi I was thinking something for just Mobile like
`pct_using_vary` 2020, `pct_of_vary_with_cache_control` 2020, `pct_using_vary` 2021, `pct_of_vary_with_cache_contro`l 2021

Which can help visualize this sentence.

```
It's important to point out that only 46.25% of total requests audited use the Vary header, but when compared to 2020, we see an overall increase by 2.85%.

Of these requests using the `Vary` header, 83.4% also have the `cache-control`. This shows us a 2.1% improvement from 2020.
```
what you think?

Reply:
I've added a few figures to this section:

- a bar chart showing the vary header directives
- two big numbers representing the 46.25% and 83.4% stats discussed below